workspaces/src/types.rs

```diff
@@ -0,0 +1,243 @@
+
+/// Signer that keeps secret key in memory.
+#[derive(Clone)]
+pub struct InMemorySigner {
+    pub account_id: AccountId,
+    pub public_key: PublicKey,
+    pub secret_key: SecretKey,
+}
+
+impl InMemorySigner {
+    pub(crate) fn from_seed(account_id: AccountId, key_type: KeyType, seed: &str) -> Self {
+        let secret_key = SecretKey::from_seed(key_type, seed);
+        Self {
+            account_id,
+            public_key: secret_key.public_key(),
+            secret_key,
+        }
+    }
+
+    pub fn from_secret_key(account_id: AccountId, secret_key: &str) -> Self {
+        let secret_key = SecretKey::from_str(secret_key).expect("Invalid secret key");
+
+        Self {
+            account_id,
+            public_key: secret_key.public_key(),
+            secret_key,
+        }
+    }
+
+    pub fn from_file(path: &Path) -> Self {
+        KeyFile::from_file(path).into()
+    }
+}
+
+impl Signer for InMemorySigner {
+    fn public_key(&self) -> PublicKey {
+        self.public_key.clone()
+    }
+
+    fn sign(&self, data: &[u8]) -> Signature {
+        self.secret_key.sign(data)
+    }
+
+    fn compute_vrf_with_proof(&self, data: &[u8]) -> (Value, Proof) {
+        let secret_key = convert_secret_key(self.secret_key.unwrap_as_ed25519());
+        secret_key.compute_vrf_with_proof(&data)
+    }
+
+    fn write_to_file(&self, path: &Path) {
+        KeyFile::from(self).write_to_file(path);
+    }
+}
+
+impl From<&InMemorySigner> for KeyFile {
+    fn from(signer: &InMemorySigner) -> KeyFile {
+        KeyFile {
+            account_id: signer.account_id.clone(),
+            public_key: signer.public_key.clone(),
+            secret_key: signer.secret_key.clone(),
+        }
+    }
+}
+
+impl From<KeyFile> for InMemorySigner {
+    fn from(key_file: KeyFile) -> Self {
+        Self {
+            account_id: key_file.account_id,
+            public_key: key_file.public_key,
+            secret_key: key_file.secret_key,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct KeyFile {
+    pub account_id: AccountId,
+    pub public_key: PublicKey,
+    pub secret_key: SecretKey,
+}
+
+impl KeyFile {
+    pub fn write_to_file(&self, path: &Path) {
+        let mut file = File::create(path).expect("Failed to create / write a key file.");
+
+        #[cfg(unix)]
+        {
+            use std::os::unix::prelude::PermissionsExt;
+            let mut perm = file
+                .metadata()
+                .expect("Failed to retrieve key file metadata.")
+                .permissions();
+
+            #[cfg(target_os = "macos")]
+            perm.set_mode(u32::from(libc::S_IWUSR | libc::S_IRUSR));
+            #[cfg(not(target_os = "macos"))]
+            perm.set_mode(libc::S_IWUSR | libc::S_IRUSR);
+
+            file.set_permissions(perm)
+                .expect("Failed to set permissions for a key file.");
+        }
+
+        let str = serde_json::to_string_pretty(self).expect("Error serializing the key file.");
```

Comment:
prob want to avoid using `str` as the keyname since it's part of Rust prelude. Didn't even know this let you

Reply:
Things I wish I didn't know part 92. 

Rust has namespaces, so it's OK to have values and types named the same: 

```rust
type X = ();
const X: X = ();
```

**but** the primitive types are doubly cursed, and have special-casing in the compiler. The `str` type and `std::str` module both live in the type namespace, but compiler still is able to disambiguate between them: 

```rust
use std::str;

fn main() {
    let assoc_fn = str::as_bytes;
    let fn_in_module = str::from_utf8;
}
```